[TOPI][x86] Introduce schedule_injective_from_existing and unify external schedules for all targets

soiferj opened the pull request:
Currently x86 `schedule_extern` does not work properly, and will treat extern ops as injective ops. This PR introduces a new generic function, `schedule_injective_from_existing` that has the core logic of `schedule_injective` for each target. `schedule_extern` then calls this method. This ends up fixing `schedule_extern` for many targets besides just x86.

Related to the discussion [here](https://discuss.tvm.ai/t/relay-register-op-pattern-based-on-target/3607/8).

@masahi @vinx13 would you be able to take a look?

vinx13 commented on topi/include/topi/x86/extern.h:
```diff
@@ -0,0 +1,89 @@
+*
+* \return A schedule for the op.
+*/
+inline Schedule schedule_extern(const Target& target, Array<Tensor> outs) {
```
is this function registered in topi.cc?

soiferj replied:
Updated

soiferj commented on topi/python/topi/generic/extern.py:
```diff
@@ -35,8 +35,12 @@ def schedule_extern(outs):
+    for out in outs:
+        if isinstance(out.op, tvm.tensor.ExternOp):
+            continue
+        _schedule_injective(out.op, s)
```
@vinx13, I have moved this logic from `cuda/extern.py` to `generic/extern.py`. Will `_schedule_injective` still call the correct overridden function per target? Or will this just call the default one?

If it just calls the default one, it seems like I have to add a new file, `x86/extern.py`

masahi commented on topi/include/topi/x86/extern.h:
```diff
@@ -0,0 +1,89 @@
+ *
+ * \return The schedule given by sch
+ */
+inline Schedule ScheduleOutputForExtern(Target target, Operation op, Schedule sch) {
```
can we use schedule_injective(...) defined in topi/x86/injective.h?

masahi commented:
@soiferj if you change the default behavior of schedule_extern in topi/python/topi/generic/extern.py, you should update topi/include/topi/generic/extern.h too.

 I would rather move your implementation of schedule_extern in x86/extern.h to generic/extern.h, and use it from both python and x86 cpp.

soiferj commented:
Sure, I can work on that

masahi commented:
Following might be useful. Look for cpp.generic or cpp.nn
https://github.com/dmlc/tvm/blob/master/topi/python/topi/generic/vision.py#L53
https://github.com/dmlc/tvm/blob/master/topi/python/topi/nn/pooling.py#L114

soiferj commented:
`x86/injective.h` and `generic/injective.h` are different. If I move the logic into `generic/extern.h`, how do I make sure I'm calling the right injective function?

masahi commented:
I see, you are right about the lack of target dispatch mechanism in our c++ topi (introduced for python in #556). It seems such dynamism is not possible in c++ topi at the moment.

I can think of two options: 
* Keep generic/extern.py and generic/extern.h as is, and duplicate existing logic in cuda/extern.py to x86/extern.py or x86/extern.h

* We make our desired change to generic/extern.py and remove generic/extern.h . It requires many code change, but it makes our codebase more consistent. For example, the cpp schedule_extern is used [here](https://github.com/dmlc/tvm/blob/master/topi/include/topi/cuda/dense.h#L92), but this is not correct. It should call topi::cuda::schedule_extern(...) directly. This bug is a result of line-by-line porting of python topi to c++.

soiferj commented:
Personally, I feel that the more logic we put in C++, the better. That way, we only have to implement things once, and users can use the Python or C++ API with the exact same features.

How about this: I'll implement this change like your first suggestion (duplicate logic in `x86/extern.h`) because this bug should be fixed urgently. 

I'll also post an RFC on the forum about adding a target dispatch API for C++. In fact, the schedules are already registered as generic functions in C++ [here](https://github.com/dmlc/tvm/blob/88cd1b1c545797eab8cb0ec32f10fc17cfef9142/topi/src/topi.cc#L714). If have an API like `topi::generic::dispath("schedule_injective")`, we can call this in C++ and always have the right implementation per target.

What do you think?

vinx13 commented:
@soiferj I agree that putting more things to C++ site is better. Lack of target dispatch in C++ can be confusing to users as they only need to call `topi::generic::...`. It would be helpful to support that in C++.

masahi commented:
@soiferj Great! I thought implementing the target dispatch in C++ would not be straightforward, but if you are willing to do it I'm happy to help. You can continue with this PR as you see fit and I'll merge this ASAP.

soiferj commented on topi/include/topi/generic/extern.h:
```diff
@@ -47,6 +48,16 @@ inline Schedule schedule_extern(const Target& target, Array<Tensor> outs) {
+      continue;
+    }
+    Array<Tensor> new_outs = { out };
+    tvm::GenericFunc::Get("schedule_injective")(new_outs);
```
@masahi or @vinx13 , this call seems to work as expected. It calls the correct schedule_injective function for the current target. However, the function that it calls creates its own schedule (see cuda/injective.py). This causes failures in the unit tests. This is probably why the previous implementation used a helper function, _schedule_injective.py). Can you think of any way to fix this when calling from C++?

vinx13 replied:
What's the problem with `cuda/injective.py`? Is overriding native generic from Python side a problem in your case?

soiferj replied:
The specific error is `Direct host side access to device memory is detected in fused_nn_conv2d_multiply_add_nn_relu_1. Did you forget to bind?`. This is being hit in `tutorials/frontend/using_external_lib.py`

soiferj replied:
I think it has to do with the fact that `schedule_injective` creates its own schedule, but I'm not totally sure.

Thanks a lot for all of the help, btw.

vinx13 replied:
You are right, we can't create a new schedule using `schedule_injective`. So the problem is that we don't want to duplicate the helper function in C++ right?

soiferj replied:
Yes, exactly. Maybe we should create a new generic function? Or have a generic function that is overridden with more arguments? Is that possible?

vinx13 replied:
I don't think generic func can be overridden by number of args. We can register the function with `TVM_REGISTER_GLOBAL`

soiferj commented:
I just added a new generic function `schedule_injective_from_existing`. I tried to update all of the core scheduling logic for all targets into this function, and changed all of the places that used `_schedule_injective` to call this function instead. Let me know what you think.

masahi commented on topi/python/topi/generic/extern.py:
```diff
@@ -35,8 +36,5 @@ def schedule_extern(outs):
-    outs = [outs] if isinstance(outs, tvm.tensor.Tensor) else outs
-    return tvm.create_schedule([x.op for x in outs])
+    target = tvm.target.current_target()
+    return cpp.generic.schedule_extern(target, outs)
```
Do we need to pass "target" variable to c++ schedule_extern?

masahi replied:
hey @soiferj can you address this comment ...

soiferj replied:
Answered below

masahi commented on topi/include/topi/cuda/injective.h:
```diff
@@ -33,21 +33,26 @@ namespace topi {
+                                                 Schedule sch,
+                                                 const Tensor& out) {
+  auto fused = detail::Fuse(sch[out], sch[out]->op.as<ComputeOpNode>()->axis);
   auto num_thread = target->max_num_threads;
```
Can be replaced by  
```
auto target = Target::Current(false);
auto num_thread = target->max_num_threads;
```

soiferj replied:
I originally tried to do this, but there was a GPU unit test that was failing :( It seemed that there wasn't a global target set, rather it was passed directly in `schedule_injective`.

This is probably also why all other generic `schedule_*` functions take the target as an argument.

masahi replied:
I see. I think this can be fixed by adding "with tvm.target.create(device)", but may be for another PR.

masahi commented:
@soiferj Do we need to pass "target" to C++ schedule_injective_from_existing? The only place it is used is inside cuda schedule_injective, and there is an alternative as I commented above. The python schedule_injective_from_existing doesn't take "target".

masahi commented:
@soiferj Also can you update your PR tile and description to better reflect what this PR is about? For people not engaged in PR and TVM monthly.

masahi commented on topi/include/topi/generic/extern.h:
```diff
@@ -47,6 +48,15 @@ inline Schedule schedule_extern(const Target& target, Array<Tensor> outs) {
+    if (out->op->derived_from<ExternOpNode>()) {
+      continue;
+    }
+    tvm::GenericFunc::Get("schedule_injective_from_existing")(s, out);
```
Since you don't pass target here, this looks like calling python-registered function ... Is that intended? (Correct me if I am wrong)

soiferj commented:
Yeah, I am little confused by that too. I tried to mirror what `schedule_injective` does. The C++ code takes in a target, and the Python code doesn’t.

In my testing, yes, the C++ code calls back into Python. I can try to remove target from the C++ interface and fix that unit test. Is that alright?

In the future, we should also remove target from the schedule function signature.

masahi commented:
Yes, I prefer fixing the unit test and cleaning up the interface.

soiferj commented:
Thanks a lot. Sorry for the long back-and-forth. I’m still pretty new to this codebase and am trying to make sure I do this the right way.

soiferj commented:
This is a CPP unit test - how can I set the current target?

soiferj commented:
@masahi or @vinx13 I am testing this new change, and while my code is being called as expected, I'm still confused as to whether this actually works. For example, I am testing the last few ops of BERT base, `matmul -> bias_add -> tanh`. In TVM, this will be fused to `fused_nn.dense_add_tanh`. When I print the `outs` that are passed into `schedule_extern`, the list has only one element, and is `tanh`. Because of this, `schedule_extern` always calls `schedule_injective_from_existing`. It seems like `nn.dense` isn't being properly scheduled.

Would one of you be able to look at this script and verify whether the behavior is expected?

`import tensorflow as tf
import tvm
import numpy as np
from tvm import relay
from tvm.contrib import graph_runtime

with tf.Graph().as_default() as graph:
    x = tf.placeholder(tf.float32, shape=(1,10))
    y = tf.placeholder(tf.float32, shape=(10,10))
    z = tf.constant(np.random.rand(10,), dtype=tf.float32)
    matmul = tf.matmul(x, y)
    add = tf.nn.bias_add(matmul, z)
    tanh = tf.math.tanh(add, name="out")
    graph_def = graph.as_graph_def(add_shapes=True)

    sym, params = relay.frontend.from_tensorflow(graph_def, shape={"Placeholder" : (1,10), "Placeholder_1" : (10,10)})
    graph, lib, params = relay.build(sym, target="llvm -mcpu=core-avx2 -libs=cblas", target_host="llvm", params = params)
    m = graph_runtime.create(graph, lib, tvm.cpu())
    m.set_input("Placeholder", np.random.rand(1,10))
    m.set_input("Placeholder_1", np.random.rand(10,10))
    m.set_input(**params)
    m.run()
    tvm_output=m.get_output(0)`

vinx13 commented:
@soiferj Only the final output of the fused group will be passed to the schedule function, and that's why we need to use `traverse` during scheduling (for example https://github.com/dmlc/tvm/blob/master/topi/python/topi/cuda/conv2d.py#L153)

soiferj commented:
I see. But if you look at line 140, extern schedule returns immediately. In my example, the fused op is never fully traversed since it calls `schedule_dense`, which immediately returns `schedule_injective` for `tanh`. There is no traversal. Should a traversal be happening within schedule_extern?

vinx13 commented:
> This is a CPP unit test - how can I set the current target?

`Target::Current`

vinx13 commented:
For your case, I'm expecting that `schedule_injective_for_existing` called for `tanh`, and `add` inlined (via `AutoInlineInjective`

soiferj commented:
Thanks, it seems you're right. AutoInlineInjective seems to do the right thing here. Also, regarding the CPP unit test, the current target can be gotten by using `Target::Current`, but it seems it cannot be set. This unit test calls `topi::cuda::schedule_injective` directly, so a target needs to be set.

vinx13 commented:
Let's fix the test. As schedules are target-specific, we expect target to be set before calling it. We can the target (`EnterTargetScope`) at the beginning of the test

soiferj commented:
Ok got the tests working :)

masahi commented on topi/src/topi.cc:
```diff
@@ -752,6 +761,30 @@ TVM_REGISTER_GENERIC_FUNC(schedule_binary_dense)
+ */
+inline PackedFunc WrapScheduleFromExisting(FTVMScheduleFromExistingBuilder builder) {
+  return PackedFunc([builder](TVMArgs args, TVMRetValue* ret) {
+    *ret = builder(args[1], args[2]);
```
Shouldn't it be args0 and args1?

soiferj replied:
You're right, fixed

masahi commented:
Great @soiferj I will merge after CI.

soiferj commented:
Awesome, thanks again for all of your help!

masahi commented:
thanks @soiferj @vinx13 this is merged.
